backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/IModularContractTypeHandler.java

```diff
@@ -0,0 +1,79 @@
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.contracts.modular;
+
+import de.metas.contracts.FlatrateTermId;
+import de.metas.contracts.modular.log.LogEntryCreateRequest;
+import de.metas.contracts.modular.log.LogEntryDeleteRequest;
+import de.metas.contracts.modular.log.LogEntryReverseRequest;
+import lombok.NonNull;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * Implementors should be annotated with {@link org.springframework.stereotype.Component},
+ * so that one instance per class is initialized and injected into {@link ModularContractService}.
+ * <p>
+ * At this stage I think there will be one implementation for {@link org.compiere.model.I_C_Order},
+ * one for {@link de.metas.inout.model.I_M_InOut} and so on.
+ * When implementing another handler, please be sure to also add a model interceptor such as {@link de.metas.contracts.modular.interceptor.C_Order}.
+ */
+public interface IModularContractTypeHandler<T>
+{
+	/**
+	 * Implementations of this method should be very fast. It's OK to return {@code true} even if not 100% sure that there will be a log record coming out of this.
+	 */
+	boolean probablyAppliesTo(@NonNull Object model);
+
+	/**
+	 * Return a request if the framework shall create the log, or {@link Optional#empty()} otherwise.
+	 */
+	@NonNull
+	Optional<LogEntryCreateRequest> createLogEntryCreateRequest(@NonNull final T model, @NonNull final FlatrateTermId flatrateTermId);
+
+	/**
+	 * Return a Stream of requests if the framework shall create one or more reversal-record log records, or {@link Optional#empty()} otherwise.
+	 * This method has no {@code settings} parameter because i *think* there are existing log records that have to be reversed independently of settings.
+	 */
+	@NonNull
+	Optional<LogEntryReverseRequest> createLogEntryReverseRequest(@NonNull final T model, @NonNull final FlatrateTermId flatrateTermId);
+
+	/**
+	 * Return a request if the framework shall delete one or more logs, or {@link Optional#empty()}  otherwise.
+	 * This method has no {@code settings} parameter because i *think* there are existing log records that have to be deleted independently of settings.
+	 */
+	@NonNull
+	Optional<LogEntryDeleteRequest> createLogEntryDeleteRequest(@NonNull final T model, @NonNull final FlatrateTermId flatrateTermId);
```

Comment:
haven't we decided against deletions from the log?

Reply:
This was in the case of a reactivation of the document. In that case, it may be ok to delete the document. But yes, I checked the PO specs, and it will not be possible to reactivate it. Removed.